wdl-cli/src/eval.rs

```diff
@@ -0,0 +1,150 @@
+
+    /// Runs a WDL task or workflow evaluation.
+    pub async fn run<R, P>(
+        mut self,
+        token: CancellationToken,
+        progress: P,
+    ) -> Result<EvaluationResult>
+    where
+        R: Future<Output = ()> + Send,
+        P: Fn(ProgressKind<'_>) -> R + Send + Sync + 'static,
+    {
+        let result = match self.inputs {
+            Inputs::Task(ref mut inputs) => {
+                // Make any paths specified in the inputs absolute.
+                let task = self.document.task_by_name(self.name).ok_or_else(|| {
+                    anyhow!(
+                        "document does not contain a task named `{name}`",
+                        name = self.name
+                    )
+                })?;
+
+                // Ensure all the paths specified in the inputs are relative to
+                // their respective origin paths.
+                inputs.join_paths(task, |key| {
+                    self.origins
+                        .get(key)
+                        .map(|p| p.to_path_buf())
+                        .ok_or(anyhow!("unable to find origin path for key `{key}`"))
+                })?;
+
+                let evaluator = TaskEvaluator::new(self.config, token).await?;
+
+                evaluator
+                    .evaluate(self.document, task, &inputs, self.output_dir, progress)
+                    .await
+                    .and_then(EvaluatedTask::into_result)
+            }
+            Inputs::Workflow(mut inputs) => {
+                let workflow = self
+                    .document
+                    .workflow()
+                    .ok_or_else(|| anyhow!("document does not contain a workflow"))?;
+
+                if workflow.name() != self.name {
+                    bail!(
+                        "document does not contain a workflow named `{name}`",
+                        name = self.name
+                    );
+                }
+
+                // Ensure all the paths specified in the inputs are relative to
+                // their respective origin paths.
+                inputs.join_paths(workflow, |key| {
+                    self.origins
+                        .get(key)
+                        .map(|p| p.to_path_buf())
+                        .ok_or(anyhow!("unable to find origin path for key `{key}`"))
+                })?;
+
+                let evaluator = WorkflowEvaluator::new(self.config, token).await?;
+                evaluator
+                    .evaluate(self.document, inputs, self.output_dir, progress)
+                    .await
+            }
+        };
+
+        match result {
+            Ok(outputs) => Ok(EvaluationResult::Success(outputs)),
+            Err(e) => match e {
+                EvaluationError::Source(diagnostic) => {
+                    let source = self.document.root().text().to_string();
+                    emit_diagnostics(&self.document.path(), &source, &[diagnostic])?;
+                    Ok(EvaluationResult::Failed)
```

Comment:
I wonder if this is really needed? Can we get away with just returning the error and eliminating `EvaluationResult`?